### PR TITLE
Added magnetometer value based on location

### PR DESCRIPTION
**Pull request by ahcorde**
# 🦟 Bug fix

## Summary

The magnetometer sensor is publishing a value but it's not based on the location. This PR calculate the value based on the location.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/systems/magnetometer/Magnetometer.cc [hunk 1]
```diff
@@ -286,6 +417,34 @@ void MagnetometerPrivate::Update(
...
+            gzwarn << "Failed to update NavSat sensor enity [" << _entity
+                    << "]. Spherical coordinates not set." << std::endl;
+            return true;
+          }
+
+          auto lat_rad = GZ_DTOR(latLonEle.value().X());
+          auto lon_rad = GZ_DTOR(latLonEle.value().Y());
+
+          // Magnetic declination and inclination (radians)
+          float declination_rad = get_mag_declination(lat_rad * 180 / M_PI, lon_rad * 180 / M_PI) * M_PI / 180;
```
Comment (mjcarroll) [inline src/systems/magnetometer/Magnetometer.cc:434]: `GZ_RTOD`

File: src/systems/magnetometer/Magnetometer.cc [hunk 2]
```diff
@@ -47,6 +47,73 @@ using namespace gz;
 using namespace sim;
 using namespace systems;
 
+/** set this always to the sampling in degrees for the table below */
+#define SAMPLING_RES		10.0f
```
Comment (mjcarroll) [inline src/systems/magnetometer/Magnetometer.cc:51]: Prefer `static constexpr` I think for constants.

**Comment by mjcarroll** [comment]
I think that we need some documentation on the derivation of these tables.  It's worth-while to note that these are for a) earth only and b) only valid in a certain window (I think they are updated every 5 years).

A more generic solution would allow for the consumption of the world magnetic model: https://www.ncei.noaa.gov/products/world-magnetic-model, but I understand if that's out of the scope of the current PR.

**Comment by ahcorde** [comment]
I used this code from this other Gazebo classic plugin https://github.com/PX4/PX4-SITL_gazebo-classic/blob/5610c3fb441a2f3babc8ad7a63c8c4ce3e40abfa/src/geo_mag_declination.cpp

**Comment by mjcarroll** [comment]
> I used this code from this other Gazebo classic plugin

Ah, makes sense, all good then.  I think the only other thing I would request is a switch to be able to switch between a constant magnetic declination/field and a varying one.  That way people can opt into the new behavior (in the unlikely case someone was depending on a constant field?)

**Comment by ahcorde** [comment]
> > I used this code from this other Gazebo classic plugin
> 
> Ah, makes sense, all good then. I think the only other thing I would request is a switch to be able to switch between a constant magnetic declination/field and a varying one. That way people can opt into the new behavior (in the unlikely case someone was depending on a constant field?)

the spherical coordinates should turn on and off this feature.

**Comment by mjcarroll** [comment]
> the spherical coordinates should turn on and off this feature.

Perfect, I think that is reasonable.

**Review by mjcarroll — APPROVED** [review]
LGTM with green ci

**Comment by mjcarroll** [comment]
CI Failures here aren't from this PR.

**Comment by peci1** [comment]
> the spherical coordinates should turn on and off this feature.

https://github.com/gazebosim/gz-sim/blob/8515827c5730501da03372d8e0a15615ef932cf2/src/systems/magnetometer/Magnetometer.cc#L459-L465

So it is true that missing spherical coords do disable this feature, but they would probably flood the console with these warnings... Or am I missing something?

**Pull request merged**